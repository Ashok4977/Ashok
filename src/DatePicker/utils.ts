import deepmerge from 'deepmerge'
import {DatePickerConfiguration, isMultiSelection, isRangeSelection, Selection} from './types'
import {defaultConfiguration} from './DatePickerProvider'

export const sanitizeDate = (date: Date | string) => new Date(new Date(date).toDateString())

export const getInitialFocusDate = (selection?: Selection) => {
  if (!selection) return sanitizeDate(new Date())

  if (selection instanceof Date) {
    return sanitizeDate(selection)
  } else if (isMultiSelection(selection)) {
    return sanitizeDate(selection[0])
  } else if (isRangeSelection(selection)) {
    return sanitizeDate(selection.from)
  } else {
    return sanitizeDate(new Date())
  }
}

export const initializeConfigurations = (externalConfig?: DatePickerConfiguration) => {
  let sanitizedConfig = {}
  if (externalConfig) {
    sanitizedConfig = Object.entries(externalConfig).filter(item => item[1] !== undefined && item[1] !== null)
    for (const item of sanitizedConfig) {
      if (item instanceof Date) {
        sanitizeDate(item)
      }
    }
  }
  return deepmerge(defaultConfiguration, sanitizedConfig)
}
