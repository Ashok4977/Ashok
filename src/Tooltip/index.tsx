/**
 * Add animations & noDelay
 * wrap
 * add align?
 */

import React from 'react'
import {useSSRSafeId} from '@react-aria/ssr'
import type {AnchorPosition, AnchorSide, AnchorAlignment} from '@primer/behaviors'
import Box from '../Box'
import {useAnchoredPosition} from '../hooks'
import {SxProp, merge, BetterSystemStyleObject} from '../sx'

type TooltipDirection = 'nw' | 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w'
type TooltipAlign = 'left' | 'right'
export type TooltipProps = {
  /** The text content of the tooltip. This should be brief and no longer than a sentence. */
  text: string
  /** @deprecated Use `text` instead */
  'aria-label'?: string
  /** Position relative to target */
  direction?: TooltipDirection
  /** Position relative to target */
  align?: TooltipAlign
  /** Use aria-describedby or aria-labelledby */
  type?: 'description' | 'label'
  /** Tooltip target, single element */
  children: React.ReactElement
} & SxProp

// map tooltip direction to anchoredPosition props
const directionToPosition: Record<TooltipDirection, {side: AnchorSide; align: AnchorAlignment}> = {
  nw: {side: 'outside-top', align: 'start'},
  n: {side: 'outside-top', align: 'center'},
  ne: {side: 'outside-top', align: 'end'},
  e: {side: 'outside-right', align: 'center'},
  se: {side: 'outside-bottom', align: 'end'},
  s: {side: 'outside-bottom', align: 'center'},
  sw: {side: 'outside-bottom', align: 'start'},
  w: {side: 'outside-left', align: 'center'}
}

// map align to AnchorAlignment
const alignToAnchorAlignment: Record<TooltipAlign, AnchorAlignment> = {left: 'start', right: 'end'}

export const TooltipContext = React.createContext<{tooltipId?: string}>({})

export const Tooltip: React.FC<TooltipProps> = ({text, children, direction = 'n', align, type = 'description'}) => {
  const tooltipRef = React.useRef<HTMLDivElement>(null)
  const anchorElementRef = React.useRef<HTMLElement>(null)

  const {position} = useAnchoredPosition({
    side: directionToPosition[direction].side,
    // maintain backward compatibility
    align: align ? alignToAnchorAlignment[align] : directionToPosition[direction].align,
    floatingElementRef: tooltipRef,
    anchorElementRef
  })

  const tooltipId = useSSRSafeId()
  const child = React.cloneElement(React.Children.only(children), {
    ref: anchorElementRef,
    [type === 'description' ? 'aria-describedby' : 'aria-labelledby']: tooltipId
  })

  return (
    <Box
      as="span"
      sx={{
        ':hover, :focus-within': {'[data-component=tooltip]': {visibility: 'visible'}}
      }}
    >
      <TooltipContext.Provider value={{tooltipId}}>{child}</TooltipContext.Provider>
      <FloatingTooltip id={tooltipId} ref={tooltipRef} position={position}>
        {text}
      </FloatingTooltip>
    </Box>
  )
}

const FloatingTooltip = React.forwardRef<
  HTMLDivElement,
  {id: string; children: string; position?: AnchorPosition} & SxProp
>(({id, children, position, sx = {}}, ref) => {
  const styles = {
    visibility: 'hidden',
    backgroundColor: 'neutral.emphasisPlus',
    color: 'fg.onEmphasis',
    borderRadius: 1,
    fontSize: 0,
    paddingY: 1,
    paddingX: 2,
    width: 'fit-content',
    maxWidth: '250px',
    textAlign: 'center',
    position: 'absolute',
    zIndex: 2,
    top: position?.top,
    left: position?.left
  }

  return (
    <Box
      role="tooltip"
      id={id}
      aria-hidden
      data-component="tooltip"
      ref={ref}
      sx={merge<BetterSystemStyleObject>(styles, sx)}
    >
      <Box
        sx={{
          position: 'relative',
          ':before': {
            content: '""',
            width: 0,
            height: 0,
            borderLeft: '5px solid transparent',
            borderRight: '5px solid transparent',
            borderBottom: '5px solid',
            borderBottomColor: 'neutral.emphasisPlus',
            position: 'absolute',
            top: '-8px',
            left: {
              start: 0,
              center: 'calc(50% - 8px)',
              end: 'calc(100% - 8px)'
            }[position?.anchorAlign || 'center']
          }
        }}
      />
      {children}
    </Box>
  )
})
