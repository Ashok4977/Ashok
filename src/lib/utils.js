
/**
 * Splice items from an Array for with the test function returns
 * truthy, modifying the original list and returning an array of
 * the spliced items.
 *
 * ```js
 * const children = React.Children.toArray(props.children)
 * const buttons = splice(children, child => child.type === Button)
 * ```
 */
const splice = (list, test) => {
  return list.reduce((acc, item, i) => {
    if (test(item)) {
      return acc.concat(list.splice(i, 1))
    }
    return acc
  }, [])
}

export {splice}
