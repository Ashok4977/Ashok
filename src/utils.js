import chroma from 'chroma-js'
import {css} from 'styled-components'
import {theme} from 'styled-system'

// color shortcuts
export const color = name => theme(`colors.${name}`)
export const black = color('black')
export const white = color('white')

/**
 * Reduce a value to a primitive as long as it is a function.
 *
 * interpolate(() => 'x') // returns 'x'
 * interpolate(() => () => 'y') // returns 'y'
 */
export const interpolate = (value, context, props) => {
  while (typeof value === 'function') {
    value = value.call(context, props)
  }
  return value
}

/**
 * This decorator interpolates all of the function's arguments by calling any
 * functions in the provided context and with the provided arguments. In other
 * words, the function you provide is guaranteed to have all of its arguments
 * "called" if they are functions. Example:
 *
 * const rgba = interpolator((rgb, a) => chroma(rgb).alpha(a).css())
 * const fadeed = rgba(theme('colors.black'), .5) // returns a function!
 * faded({}) // probably throws an error because props.theme isn't set
 */
export const interpolator = f => {
  return (...outer) => {
    return function(props) {
      return f(...[...outer].map(a => interpolate(a, this, props)))
    }
  }
}

export const rgba = interpolator(       (rgb, alpha)    => chroma(rgb).alpha(alpha).css())
export const lighten = interpolator(    (value, amount) => chroma(value).brighten(amount).css())
export const darken = interpolator(     (value, amount) => chroma(value).darken(amount).css())
export const desaturate = interpolator( (value, amount) => chroma(value).desaturate(amount).css())
export const mix = interpolator(        (a, b, amount)  => chroma.mix(a, b, amount).css())
