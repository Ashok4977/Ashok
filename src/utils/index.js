
export function defined(val) {
  return val !== null && typeof val !== 'undefined'
}

export function definedFallback(input, fallback) {
  return defined(input) ? input : fallback
}
