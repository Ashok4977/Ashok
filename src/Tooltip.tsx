import React from 'react'
import {isSupported, apply} from '@oddbird/popover-polyfill/fn'
import styled from 'styled-components'
import {get} from './constants'
import sx, {SxProp} from './sx'
import {ComponentProps} from './utils/types'
import classNames from 'classnames'

interface PopoverProps extends SxProp, React.PropsWithChildren {
  popover: 'auto'
  id: string
  'aria-label': string
}

const TooltipBase = styled('span')
  .withConfig({
    shouldForwardProp: (prop, defaultValidatorFn) => ['popover'].includes(prop) || defaultValidatorFn(prop),
  })
  .attrs<PopoverProps>(props => ({
    popover: props.popover,
    id: props.id,
  }))<PopoverProps>`
  /* Without the triple ampersands, the specificity of these styles are overridden by the base popover styles.
   * https://styled-components.com/docs/faqs#how-can-i-override-styles-with-higher-specificity
   */
  &&& {
    border: none;
    overflow: hidden;
    margin: 0;
    position: fixed;
    background: transparent;
    top: 0;
    left: 0;
    display: none;
    opacity: 0;
    padding: 0;
    --width: 0;
    --height: 0;
    --anchor-width: 0;

    @keyframes tooltip-appear {
      from {
        opacity: 0;
      }

      to {
        opacity: 1;
      }
    }

    animation-name: tooltip-appear;
    animation-duration: 0.1s;
    animation-fill-mode: forwards;
    animation-timing-function: ease-in;
    animation-delay: 0.4s;

    &.tooltipped-no-delay {
      animation-delay: 0s;
    }

    &.\\:open {
      display: block;
    }

    &:open {
      display: block;
    }

    &::before {
      position: relative;
      padding: 0.5em 0.75em;
      border-radius: ${get('radii.1')};
      font: normal normal 11px/1.5 ${get('fonts.normal')};
      background: ${get('colors.neutral.emphasisPlus')};
      color: ${get('colors.fg.onEmphasis')};
      content: attr(aria-label);
      word-wrap: break-word;
      text-align: center;
      display: block;
    }

    &::after {
      position: absolute;
      content: '';
      right: 50%;
      margin-right: -6px;
      width: 0;
      height: 0;
      border: 6px solid transparent;
    }

    &.tooltipped-n,
    &.tooltipped-ne,
    &.tooltipped-nw {
      padding-bottom: 10px;

      &::after {
        top: calc(var(--height) - 12px);
        border-top-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-nw {
      &::after {
        right: 0.5em;
      }
    }

    &.tooltipped-ne {
      &::after {
        left: 0.5em;
      }
    }

    &.tooltipped-sw {
      padding-top: 10px;
      &::after {
        right: 0.5em;
        top: 0;
        border-bottom-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-se {
      padding-top: 10px;
      &::after {
        left: 0.5em;
        top: 0;
        border-bottom-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-s {
      padding-top: 10px;

      &::after {
        top: 0;
        border-bottom-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-e {
      padding-left: 10px;

      &:after {
        top: calc(var(--height) / 2 - 5px);
        right: calc(var(--width) - 5px);
        border-right-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-w {
      padding-right: 10px;
      &::after {
        top: calc(var(--height) / 2 - 5px);
        right: 4px;
        border-left-color: ${get('colors.neutral.emphasisPlus')};
      }
    }

    &.tooltipped-multiline {
      &::before {
        width: max-content;
        max-width: 250px;
        word-wrap: break-word;
        white-space: pre-line;
        border-collapse: separate;
      }
    }
    ${sx}
  }
`

type TooltipProps = {
  direction?: 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w' | 'nw'
  text?: string
  noDelay?: boolean
  wrap?: boolean
} & ComponentProps<typeof TooltipBase>

const Tooltip = ({direction = 'n', children, className, noDelay, wrap, popover = 'auto', ...rest}: TooltipProps) => {
  if (!isSupported()) {
    apply()
  }

  const classes = classNames(
    className,
    `tooltipped-${direction}`,
    noDelay && 'tooltipped-no-delay',
    wrap && 'tooltipped-multiline',
  )

  const popoverRef = React.useRef<HTMLElement>(null)

  const [showTooltip, setShowTooltip] = React.useState(false)

  const handlePointerEnter = (event: React.PointerEvent<HTMLElement> | React.FocusEvent<HTMLElement>) => {
    const anchored = popoverRef.current
    if (anchored !== null && !showTooltip) {
      anchored.showPopover()
      anchored.style.setProperty('--width', `${anchored.clientWidth}px`)
      anchored.style.setProperty('--height', `${anchored.clientHeight}px`)

      const anchor = event.currentTarget as HTMLElement
      const {top, left, width, height} = anchor.getBoundingClientRect()
      anchored.style.setProperty('--anchor-width', `${width}px`)
      let [_top, _left] = [0, 0]

      if (direction === 'n' || direction === 'ne' || direction === 'nw') {
        _top = top - anchored.clientHeight
        _left = left - anchored.clientWidth / 2 + width / 2
      }

      if (direction === 's' || direction === 'sw' || direction === 'se') {
        _top = top + anchored.clientHeight / 2
        _left = left - anchored.clientWidth / 2 + width / 2
      }

      if (direction === 'e' || direction === 'w') {
        _top = top - (anchored.clientHeight - height) / 2
      }

      if (direction === 'nw' || direction === 'sw') {
        _left = left + width - anchored.clientWidth
      }

      if (direction === 'w') {
        _left = left - anchored.clientWidth
      }

      if (direction === 'se' || direction === 'ne') {
        _left = left
      }

      if (direction === 'e') {
        _left = left + width
      }

      anchored.style.setProperty('top', `${_top}px`)
      anchored.style.setProperty('left', `${_left}px`)
      setShowTooltip(true)
    }
  }

  const handlePointerLeave = () => {
    const anchored = popoverRef.current
    if (anchored !== null && showTooltip) {
      anchored.hidePopover()
      setShowTooltip(false)
    }
  }

  return (
    <>
      <span
        style={{display: 'inline-block'}}
        onPointerEnter={handlePointerEnter}
        onPointerLeave={handlePointerLeave}
        onFocus={handlePointerEnter}
        onBlur={handlePointerLeave}
      >
        {children}
      </span>
      <TooltipBase ref={popoverRef} popover={popover} className={classes} {...rest} />
    </>
  )
}

export default Tooltip
