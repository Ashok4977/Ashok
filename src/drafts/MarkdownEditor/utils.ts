import {isMacOS} from '@primer/behaviors/utils'

export type InputState = {
  readonly value: string
  readonly selectionStart: number
  readonly selectionEnd: number
}

export const getSelectedLineRange = (inputState: InputState): [number, number] => {
  // Subtract one from the caret position so the newline found is not the one _at_ the caret position
  // then add one because we don't want to include the found newline. Also changes -1 (not found) result to 0
  const start = inputState.value.lastIndexOf('\n', inputState.selectionStart - 1) + 1

  // activeLineEnd will be the index of the next newline inclusive, which works because slice is last-index exclusive
  let end = inputState.value.indexOf('\n', inputState.selectionEnd)
  if (end === -1) end = inputState.value.length

  return [start, end]
}

export const markdownComment = (text: string) => `<!-- ${text.replaceAll('--', '\\-\\-')} -->`

export const markdownLink = (text: string, url: string) =>
  `[${text.replaceAll('[', '\\[').replaceAll(']', '\\]')}](${url.replaceAll('(', '\\(').replaceAll(')', '\\)')})`

export const markdownImage = (altText: string, url: string) => `!${markdownLink(altText, url)}`

export const isModifierKey = (event: KeyboardEvent | React.KeyboardEvent<unknown>) =>
  isMacOS() ? event.metaKey : event.ctrlKey

export function isMultipleLines(string: string): boolean {
  return string.trim().split('\n').length > 1
}
