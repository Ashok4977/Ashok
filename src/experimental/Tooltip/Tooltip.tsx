import React, {Children, useEffect, useRef, useState} from 'react'
import Box from '../../Box'
import sx, {SxProp} from '../../sx'
import {useId} from '../../hooks/useId'
import {invariant} from '../../utils/invariant'
import {warning} from '../../utils/warning'
import styled from 'styled-components'
import {get} from '../../constants'
import {ComponentProps} from '../../utils/types'
import {getAnchoredPosition} from '@primer/behaviors'
import type {AnchorSide, AnchorAlignment} from '@primer/behaviors'

declare global {
  interface PopoverToggleTargetElementInvoker {
    popoverTargetElement: HTMLElement | null
    popoverTargetAction: 'toggle' | 'show' | 'hide'
  }
  interface ToggleEvent extends Event {
    oldState: string
    newState: string
  }
  interface HTMLElement {
    popover: 'auto' | 'manual' | null
    showPopover(): void
    hidePopover(): void
    togglePopover(): void
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface HTMLButtonElement extends PopoverToggleTargetElementInvoker {}
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface HTMLInputElement extends PopoverToggleTargetElementInvoker {}

  interface Window {
    ToggleEvent: ToggleEvent
  }
}

const StyledTooltip = styled.div`
  // tooltip element itself
  position: relative;
  z-index: 1000000;
  padding: 0.5em 0.75em;
  font: normal normal 11px/1.5 ${get('fonts.normal')};
  -webkit-font-smoothing: subpixel-antialiased;
  color: ${get('colors.fg.onEmphasis')};
  text-align: center;
  text-decoration: none;
  text-shadow: none;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: break-word;
  white-space: normal;
  background: ${get('colors.neutral.emphasisPlus')};
  border-radius: ${get('radii.2')};
  border: 0;
  width: max-content;
  opacity: 0;
  max-width: 250px;
  inset: auto;
  /* for scrollbar */
  overflow: hidden;
  @media (forced-colors: active) {
    outline: 1px solid transparent;
  }

  /* tooltip element should be rendered visually hidden when it is not opened.  */
  &:not(:popover-open) {
    /* Visually hidden styles */
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    white-space: nowrap;
    border-width: 0;
  }
  // the caret
  /* &::before {
    position: absolute;
    z-index: 1000001;
    color: ${get('colors.neutral.emphasisPlus')};
    content: '';
    border: 6px solid transparent;
    opacity: 0;
  } */

  // This is needed to keep the tooltip open when the user leaves the trigger element to hover tooltip
  &::after {
    position: absolute;
    display: block;
    right: 0;
    left: 0;
    height: 12px;
    content: '';
  }

  // delay animation for tooltip
  @keyframes tooltip-appear {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }

  /* South, East, Southeast, Southwest before */

  &:popover-open[data-direction='n']::before,
  &[data-direction='ne']::before,
  &[data-direction='nw']::before {
    top: 82%;
    border-top-color: ${get('colors.neutral.emphasisPlus')};
  }
  &[data-direction='s']::before,
  &[data-direction='se']::before,
  &[data-direction='sw']::before {
    bottom: 100%;
    border-bottom-color: ${get('colors.neutral.emphasisPlus')};
  }
  &[data-direction='n']:before,
  &[data-direction='s']:before {
    right: 50%;
    margin-right: -6px;
  }
  &[data-direction='ne']::before,
  &[data-direction='se']::before {
    left: 0;
    margin-left: 6px;
  }
  &[data-direction='sw']::before,
  &[data-direction='nw']::before {
    right: 0;
    margin-right: 6px;
  }
  /* South, East, Southeast, Southwest after */
  &[data-direction='n']::after,
  &[data-direction='ne']::after,
  &[data-direction='nw']::after {
    top: 100%;
  }
  &[data-direction='s']::after,
  &[data-direction='se']::after,
  &[data-direction='sw']::after {
    bottom: 100%;
  }
  /* West before and after */
  &[data-direction='w']::before {
    top: 50%;
    bottom: 50%;
    left: 100%;
    margin-top: -6px;
    border-left-color: ${get('colors.neutral.emphasisPlus')};
  }
  &[data-direction='w']::after {
    position: absolute;
    display: block;
    height: 100%;
    width: 8px;
    content: '';
    bottom: 0;
    left: 100%;
  }
  /* East before and after */
  &[data-direction='e']::after {
    position: absolute;
    display: block;
    height: 100%;
    width: 8px;
    content: '';
    bottom: 0;
    right: 100%;
    margin-left: -8px;
  }
  &[data-direction='e']::before {
    top: 50%;
    bottom: 50%;
    right: 100%;
    margin-top: -6px;
    border-right-color: ${get('colors.neutral.emphasisPlus')};
  }

  /* Animation styles */

  &:popover-open,
  &:popover-open::before {
    animation-name: tooltip-appear;
    animation-duration: 0.1s;
    animation-fill-mode: forwards;
    animation-timing-function: ease-in;
    animation-delay: 0.4s;
  }

  /* Position of the tooltip element when it is opened. */

  &:popover-open {
    &[data-no-delay='true'],
    &[data-no-delay='true']::before {
      animation-delay: 0s;
    }
  }

  ${sx};
`

type TooltipDirection = 'nw' | 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w'
export type TooltipProps = React.PropsWithChildren<
  {
    direction?: TooltipDirection
    text?: string
    noDelay?: boolean
    type?: 'label' | 'description'
  } & SxProp &
    ComponentProps<typeof StyledTooltip>
>

export type TriggerPropsType = {
  'aria-describedby'?: string
  'aria-labelledby'?: string
  'aria-label'?: string
  onBlur?: React.FocusEventHandler
  onFocus?: React.FocusEventHandler
  onMouseEnter?: React.MouseEventHandler
  ref?: React.RefObject<HTMLElement>
}

// map tooltip direction to anchoredPosition props
const directionToPosition: Record<TooltipDirection, {side: AnchorSide; align: AnchorAlignment}> = {
  nw: {side: 'outside-top', align: 'start'},
  n: {side: 'outside-top', align: 'center'},
  ne: {side: 'outside-top', align: 'end'},
  e: {side: 'outside-right', align: 'center'},
  se: {side: 'outside-bottom', align: 'end'},
  s: {side: 'outside-bottom', align: 'center'},
  sw: {side: 'outside-bottom', align: 'start'},
  w: {side: 'outside-left', align: 'center'},
}

// The list is from GitHub's custom-axe-rules https://github.com/github/github/blob/master/app/assets/modules/github/axe-custom-rules.ts#L3
const interactiveElements = ['a[href]', 'button', 'summary', 'select', 'input:not([type=hidden])', 'textarea']

const isInteractive = (element: HTMLElement) => {
  return (
    interactiveElements.some(selector => element.matches(selector)) ||
    (element.hasAttribute('role') && element.getAttribute('role') === 'button')
  )
}
export const Tooltip = ({
  direction = 'n',
  // tooltip text
  text,
  type = 'description',
  noDelay,
  children,
  ...rest
}: TooltipProps) => {
  const id = useId()
  const triggerRef = useRef<HTMLElement>(null)
  const tooltipElRef = useRef<HTMLDivElement>(null)
  const child = Children.only(children)
  const [open, setOpen] = useState(false)

  // we need this check for every render
  if (__DEV__) {
    // We don't want these warnings to show up on tests because it fails the tests (at dotcom) due to not extecting a warning.
    // TODO: We can remove the test check when we update these warnings to invariants.
    // Practically, this is not a conditional hook, it is just making sure this hook runs only on DEV not PROD.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      if (triggerRef.current) {
        // Has trigger element or any of its children interactive elements?
        const isTriggerInteractive = isInteractive(triggerRef.current)
        const triggerChildren = triggerRef.current.childNodes
        const hasInteractiveChild = Array.from(triggerChildren).some(child => {
          return child instanceof HTMLElement && isInteractive(child)
        })
        invariant(
          isTriggerInteractive || hasInteractiveChild,
          'The `Tooltip` component expects a single React element that contains interactive content. Consider using a `<button>` or equivalent interactive element instead.',
        )
        // If the tooltip is used for labelling the interactive element, the trigger element or any of its children should not have aria-label
        if (type === 'label') {
          const hasAriaLabel = triggerRef.current.hasAttribute('aria-label')
          const hasAriaLabelInChildren = Array.from(triggerRef.current.childNodes).some(
            child => child instanceof HTMLElement && child.hasAttribute('aria-label'),
          )
          warning(
            hasAriaLabel || hasAriaLabelInChildren,
            'The label type `Tooltip` is going to be used here to label the trigger element. Please remove the aria-label from the trigger element.',
          )
        }
      }
    })
  }

  // Opting in using useOnEscapePress hook instead of onKeydown event to be able to close the tooltip when the mouse is hovering on the trigger element
  // useOnEscapePress(
  //   (e: KeyboardEvent) => {
  //     if (open) {
  //       e.stopPropagation()
  //       setOpen(false)
  //     }
  //   },
  //   [open],
  // )

  const openTooltip = () => {
    if (tooltipElRef.current && triggerRef.current && !tooltipElRef.current.matches(':popover-open')) {
      //
      tooltipElRef.current.showPopover()
      setOpen(true)
    }
  }

  // const position = {top, left}

  useEffect(() => {
    if (!tooltipElRef.current || !triggerRef.current) return
    const tooltip = tooltipElRef.current
    const trigger = triggerRef.current
    tooltip.setAttribute('popover', 'auto')
    // if (!open) return
    const settings = {
      side: directionToPosition[direction].side,
      align: directionToPosition[direction].align,
      // alignmentOffset: 10,
      // anchorOffset: -2,
    }

    const positionSet = () => {
      const {top, left} = getAnchoredPosition(tooltip, trigger, settings)

      tooltip.style.top = `${top}px`
      tooltip.style.left = `${left}px`
      tooltip.setAttribute('data-setting', JSON.stringify(settings))
    }

    tooltip.addEventListener('toggle', positionSet)

    return () => {
      tooltip.removeEventListener('toggle', positionSet)
    }
  }, [tooltipElRef, triggerRef, direction])

  return (
    <Box
      sx={{display: 'inline-block'}}
      onMouseLeave={() => {
        if (tooltipElRef.current?.matches(':popover-open')) tooltipElRef.current.hidePopover()
      }}
    >
      {React.isValidElement(child) &&
        React.cloneElement(child as React.ReactElement<TriggerPropsType>, {
          ref: triggerRef,
          // If it is a type description, we use tooltip to describe the trigger
          'aria-describedby': type === 'description' ? id : undefined,
          // If it is a label type, we use tooltip to label the trigger
          'aria-labelledby': type === 'label' ? id : undefined,
          onBlur: (event: React.FocusEvent) => {
            if (tooltipElRef.current?.matches(':popover-open')) tooltipElRef.current.hidePopover()
            child.props.onBlur?.(event)
          },
          onFocus: (event: React.FocusEvent) => {
            openTooltip()
            // if (!tooltipElRef.current?.matches(':popover-open')) tooltipElRef.current?.showPopover()
            child.props.onFocus?.(event)
          },
          onMouseEnter: (event: React.MouseEvent) => {
            openTooltip()
            child.props.onMouseEnter?.(event)
          },
        })}
      <StyledTooltip
        ref={tooltipElRef}
        data-direction={direction}
        data-state={open ? 'open' : undefined}
        data-no-delay={noDelay}
        {...rest}
        // Only need tooltip role if the tooltip is a description for supplementary information
        role={type === 'description' ? 'tooltip' : undefined}
        // stop AT from announcing the tooltip twice when it is a label type because it will be announced with "aria-labelledby"
        aria-hidden={type === 'label' ? true : undefined}
        id={id}
      >
        {text}
      </StyledTooltip>
    </Box>
  )
}
