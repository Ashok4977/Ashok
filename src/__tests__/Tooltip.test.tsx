import React from 'react'
import Tooltip, {TooltipProps} from '../Tooltip'
import {ReactTestRendererJSON} from 'react-test-renderer'
import {render, childRenderClasses, childRendersClass, behavesAsComponent, checkExports} from '../utils/testing'
import {render as HTMLRender} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('Tooltip', () => {
  behavesAsComponent({Component: Tooltip, options: {skipAs: true}})

  checkExports('Tooltip', {
    default: Tooltip,
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Tooltip text="hi" />)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
  })

  it('renders a <span> with the "tooltipped" class', () => {
    expect(render(<Tooltip />).type).toEqual('span')
    expect(childRenderClasses(<Tooltip />, -1)).toContain('tooltipped-n')
  })

  it('respects the "direction" prop', () => {
    for (const direction of Tooltip.directions) {
      expect(
        childRendersClass(
          <Tooltip direction={direction as TooltipProps['direction']} />,
          `tooltipped-${direction}`,
          -1,
        ),
      ).toBe(true)
    }
  })

  it('respects the "noDelay" prop', () => {
    expect(childRendersClass(<Tooltip noDelay />, 'tooltipped-no-delay', -1)).toBe(true)
  })

  it('respects the "text" prop', () => {
    const tooltip = render(<Tooltip text="hi" />) as ReactTestRendererJSON
    const child = tooltip.children?.at(-1) as ReactTestRendererJSON
    expect(child.props['aria-label']).toEqual('hi')
  })

  it('respects the "wrap" prop', () => {
    expect(childRendersClass(<Tooltip wrap />, 'tooltipped-multiline', -1)).toBe(true)
  })
})
