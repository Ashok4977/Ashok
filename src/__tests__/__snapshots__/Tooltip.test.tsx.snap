// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders consistently 1`] = `
.c0.c0.c0 {
  border: none;
  overflow: hidden;
  margin: 0;
  position: fixed;
  background: transparent;
  top: 0;
  left: 0;
  display: none;
  opacity: 0;
  padding: 0;
  --width: 0;
  --height: 0;
  --anchor-width: 0;
  -webkit-animation-name: tooltip-appear;
  animation-name: tooltip-appear;
  -webkit-animation-duration: 0.1s;
  animation-duration: 0.1s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-timing-function: ease-in;
  animation-timing-function: ease-in;
  -webkit-animation-delay: 0.4s;
  animation-delay: 0.4s;
}

.c0.c0.c0.tooltipped-no-delay {
  -webkit-animation-delay: 0s;
  animation-delay: 0s;
}

.c0.c0.c0.\\:open {
  display: block;
}

.c0.c0.c0:open {
  display: block;
}

.c0.c0.c0::before {
  position: relative;
  padding: 0.5em 0.75em;
  border-radius: 3px;
  font: normal normal 11px/1.5 -apple-system,BlinkMacSystemFont,"Segoe UI","Noto Sans",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji";
  background: #24292f;
  color: #ffffff;
  content: attr(aria-label);
  word-wrap: break-word;
  text-align: center;
  display: block;
}

.c0.c0.c0::after {
  position: absolute;
  content: '';
  right: 50%;
  margin-right: -6px;
  width: 0;
  height: 0;
  border: 6px solid transparent;
}

.c0.c0.c0.tooltipped-n,
.c0.c0.c0.tooltipped-ne,
.c0.c0.c0.tooltipped-nw {
  padding-bottom: 10px;
}

.c0.c0.c0.tooltipped-n::after,
.c0.c0.c0.tooltipped-ne::after,
.c0.c0.c0.tooltipped-nw::after {
  top: calc(var(--height) - 12px);
  border-top-color: #24292f;
}

.c0.c0.c0.tooltipped-nw::after {
  right: 0.5em;
}

.c0.c0.c0.tooltipped-ne::after {
  left: 0.5em;
}

.c0.c0.c0.tooltipped-sw {
  padding-top: 10px;
}

.c0.c0.c0.tooltipped-sw::after {
  right: 0.5em;
  top: 0;
  border-bottom-color: #24292f;
}

.c0.c0.c0.tooltipped-se {
  padding-top: 10px;
}

.c0.c0.c0.tooltipped-se::after {
  left: 0.5em;
  top: 0;
  border-bottom-color: #24292f;
}

.c0.c0.c0.tooltipped-s {
  padding-top: 10px;
}

.c0.c0.c0.tooltipped-s::after {
  top: 0;
  border-bottom-color: #24292f;
}

.c0.c0.c0.tooltipped-e {
  padding-left: 10px;
}

.c0.c0.c0.tooltipped-e:after {
  top: calc(var(--height) / 2 - 5px);
  right: calc(var(--width) - 5px);
  border-right-color: #24292f;
}

.c0.c0.c0.tooltipped-w {
  padding-right: 10px;
}

.c0.c0.c0.tooltipped-w::after {
  top: calc(var(--height) / 2 - 5px);
  right: 4px;
  border-left-color: #24292f;
}

.c0.c0.c0.tooltipped-multiline::before {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 250px;
  word-wrap: break-word;
  white-space: pre-line;
  border-collapse: separate;
}

<span
  onBlur={[Function]}
  onFocus={[Function]}
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
  style={
    {
      "display": "inline-block",
    }
  }
>
  <span
    className="c0 tooltipped-n"
    popover="auto"
  />
</span>
`;
